src/libraries/System.Net.Mail/src/System/Net/Mail/SmtpClient.cs

```diff
@@ -810,20 +828,45 @@ public Task SendMailAsync(MailMessage message)
                 throw;
             }
 
+            // Only register on the CT if HandleCompletion hasn't started to ensure the CTR is disposed
+            bool lockTaken = false;
+            try
+            {
+                Monitor.TryEnter(tcs, ref lockTaken);
```

Comment:
What necessitates synchronization here?

Reply:
We don't want to run the Cancel method if the operation already finished and the TCS is being set. 

If the request just finished as we've registered the CTR, the CTR wouldn't be disposed, potentially canceling other operations.

Reply:
While some synchronization is necessary, a lock shouldn't be.  For example, this should be doable with a basic CompareExchange, e.g.
```C#
int state = 0;
...
// when registering for cancellation
ctr = ct.Register(...);
if (Interlocked.CompareExchange(ref state, 1, 0) == 1) ctr.Dispose();
...
// when the operation completes
SendCompleted -= handler;
if (Interlocked.CompareExchange(ref state, 1, 0) == 1) ctr.Dispose();
```
If the registering thread stores the ctr before the operation completes, then it will see state as 1 and proceed to dispose the ctr, which is guaranteed to already be set because the state was transitioned to 1 after it was set.  If the operation completes before ctr is stored, then the registering thread will see state as 1 and will dispose the registration it just created; it also won't yet have handed back the Task to its caller, so there can't be any continuations from that Task yet.

That then also lets you revert the change to the TCS ctor, as you're not completing the task while holding a lock.

Reply:
```c#
// when the operation completes
SendCompleted -= handler;
if (Interlocked.CompareExchange(ref state, 1, 0) == 1) ctr.Dispose();
```
Does the CompareExchange provide a guarantee that a read of `ctr` will pick up the value written in the registering thread?

Reply:
Yes. The compare exchange is a full memory barrier.

Reply:
I like it.
I'll update the PR when I wake up

Reply:
@stephentoub Please re-review. I suppose in this case Exchange can be used instead.

Reply:
Yup, Exchange should be fine. The important part is the atomic transition and read of the prior state.